diff --git a/src/org.kde.dolphin.desktop b/src/org.kde.dolphin.desktop
index c0ac84347..b624d5a8b 100755
--- a/src/org.kde.dolphin.desktop
+++ b/src/org.kde.dolphin.desktop
@@ -44,7 +44,7 @@ Name[x-test]=xxDolphinxx
 Name[zh_CN]=Dolphin
 Name[zh_TW]=Dolphin
 Exec=dolphin %u
-Icon=system-file-manager
+Icon=org.kde.dolphin
 Type=Application
 X-DocPath=dolphin/index.html
 Categories=Qt;KDE;System;FileTools;FileManager;
